syntax = "proto3";

package sumo.engine.v1;

message KeyValue {
    string key = 1;
    oneof value {
      double double_value = 2;
      string string_value = 3;
    }
}

message TelemetryFrame {
  string run_id = 1;
  int64 step = 2;
  // double sim_time_s = 3;
  repeated KeyValue metrics = 4;
}


message TlsSetPhase {
  string tls_id = 1;
  int32 phase_index = 2;
}

message Action {
  oneof payload {
    TlsSetPhase tls_set_phase = 1;
  }
}

message ActionBundle {
  string run_id = 1;
  int64 step = 2;
  repeated Action actions = 3;
}

message Artifact {
  string artifact_id = 1;
  string type = 2;
  string uri = 3;
  string sha256 = 4;
}

message CreateRunRequest {
  string sumocfg_path = 1;
  string sumo_binary = 2;
  int32 step_length_ms = 3;
}

message CreateRunResponse {
  string run_id = 1;
  repeated Artifact input_artifacts = 2;
}

message RunRequest {
  string run_id = 1;
  int64 max_steps = 2;
}

message RunResponse {
  string run_id = 1;
}

message CloseRunRequest {
  string run_id = 1;
}

message CloseRunResponse {
  string run_id = 1;
  // maybe can return some analytics also?
}

message ApplyActionsResponse {
  string run_id = 1;
}

message StreamRequest {
  string run_id = 1;
}

message Parameter {
  string name = 1;
  string value = 2;
}

message SubscribeRequest {
  string run_id = 1;
  string domain = 2;
  string getter_name = 3;
  optional string object_id = 4;
  repeated Parameter additional_parameters = 5;
  optional string name = 6;
}

message SubscribeResponse {
  string subscription_name_or_fingerprint = 1;
}

message UnsubscribeRequest {
  string run_id = 1;
}

message UnsubscribeResponse {
  string run_id = 1;
}

service EngineService {
  rpc CreateRun(CreateRunRequest) returns (CreateRunResponse);
  rpc Run(RunRequest) returns (RunResponse);
  rpc CloseRun(CloseRunRequest) returns (CloseRunResponse);
  rpc ApplyActions(ActionBundle) returns (ApplyActionsResponse);
  rpc StreamTelemetry(StreamRequest) returns (stream TelemetryFrame);
  rpc StreamSubscriptions(StreamRequest) returns (stream TelemetryFrame);
  rpc Subscribe(SubscribeRequest) returns (SubscribeResponse);
  rpc Unsubscribe(UnsubscribeRequest) returns (UnsubscribeResponse);
}

